File: leetcode/easy/0100SameTree/sametree.cpp
#include <iostream>
#include <stack>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class RecursiveSolution
{
public:
    bool isSameTree(TreeNode *p, TreeNode *q)
    {

        if (!p && !q)
            return true;
        if (!p || !q)
            return false;
        if (p->val != q->val)
            return false;
        return isSameTree(p->left, q->left) && isSameTree(p->right, q->right);
    }
};

class IterativeSolution
{
public:
    bool isSameTree(TreeNode *p, TreeNode *q)
    {

        stack<TreeNode *> stack1, stack2;
        stack1.push(p), stack2.push(q);
        while (!stack1.empty() || !stack2.empty())
        {
            TreeNode *cur1 = stack1.top();
            stack1.pop();

            TreeNode *cur2 = stack2.top();
            stack2.pop();

            if (cur1 && !cur2)
                return false;
            if (!cur1 && cur2)
                return false;
            if (!cur1 && !cur2)
                continue;

            if (cur1->val != cur2->val)
                return false;
            stack1.push(cur1->right);
            stack1.push(cur1->left);
            stack2.push(cur2->right);
            stack2.push(cur2->left);
        }

        return stack1.empty() && stack2.empty();
    }
};

File: leetcode/easy/0100SameTree/sametree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class IterativeSolution(object):
    def isSameTree(self, p, q):
        stack1 = []
        stack2 = []
        stack1.append(p)
        stack2.append(q)

        while( len(stack1) != 0 or len(stack2) != 0):
            cur1 = stack1[-1]
            stack1.pop()
            cur2 = stack2[-1]
            stack2.pop()

            if cur1 and not cur2:
                return False
            if not cur1 and cur2:
                return False
            if not cur1 and not cur2:
                continue
            if cur1.val != cur2.val:
                return False
            stack1.append(cur1.right)
            stack1.append(cur1.left)
            stack2.append(cur2.right)
            stack2.append(cur2.left)
        
        return (len(stack1) == 0 and len(stack2) == 0)

class RecursiveSolution(object):
    def isSameTree(self, p, q):
        """
        :type p: TreeNode
        :type q: TreeNode
        :rtype: bool
        """
        if (not p and not q):
            return True
        if (not p or not q):
            return False
        if (p.val != q.val):
            return False
        return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)